File: src/types/whiteLabel.ts
/**
 * Tipos centralizados para o sistema White Label
 * 
 * Este arquivo contém todas as interfaces e tipos relacionados
 * à configuração de estilos personalizados por empresa SST.
 */

// Tipo para cores HSL
export interface HSLColor {
  h: number;
  s: number;
  l: number;
}

// Configuração de cores do White Label
export interface WhiteLabelColors {
  bgColor?: string;
  surfaceColor?: string;
  borderColor?: string;
  textColor?: string;
  mutedColor?: string;
  primaryColor?: string;
  secondaryColor?: string;
  linkColor?: string;
  iconColor?: string;
  badgeBg?: string;
  successColor?: string;
  warningColor?: string;
  errorColor?: string;
  infoColor?: string;
  buttonBg?: string;
  buttonText?: string;
  buttonHover?: string;
  buttonDisabled?: string;
  loginBg?: string;
}

// Configuração de tipografia
export interface WhiteLabelTypography {
  fontBody?: string;
  fontHeading?: string;
  baseFontSize?: number;
  fontWeight?: number;
  lineHeight?: number;
}

// Configuração de layout
export interface WhiteLabelLayout {
  density?: number;
  radius?: number;
  cardShadow?: number;
}

// Configuração de identidade/branding
export interface WhiteLabelBranding {
  title?: string;
  subtitle?: string;
  subject?: string;
  domain?: string;
  logoUrl?: string;
  faviconUrl?: string;
  loginImageUrl?: string;
  aboutText?: string;
  emailFooter?: string;
  emptyTone?: string;
}

// Configuração do Kanban
export interface WhiteLabelKanban {
  colHeaderBg?: string;
  colHeaderText?: string;
  colBorder?: string;
  colShadow?: string;
  colWidth?: number;
  colAutoWidth?: boolean;
  cardBg?: string;
  cardBorder?: string;
  cardStripe?: string;
  stripeMode?: number;
  cardCompact?: boolean;
  blockedColor?: string;
  fTitle?: boolean;
  fSubtitle?: boolean;
  fId?: boolean;
  fTags?: boolean;
  fAssignee?: boolean;
  fDate?: boolean;
  fSla?: boolean;
  fPriority?: boolean;
  fPoints?: boolean;
  fLabels?: boolean;
  labelRequired?: boolean;
  labelLimit?: number;
  labelPalette?: string;
  avatarShape?: string;
  avatarSize?: number;
  avatarPhoto?: boolean;
  aMove?: boolean;
  aDone?: boolean;
  aComment?: boolean;
  aAssign?: boolean;
}

// Configuração completa do White Label (usado no hook de aplicação)
export interface WhiteLabelConfig extends WhiteLabelColors, WhiteLabelTypography, WhiteLabelLayout {
  logoUrl?: string;
  title?: string;
  subtitle?: string;
  [key: string]: any;
}

// Configuração completa para o formulário de edição
export interface WhiteLabelFormConfig extends 
  WhiteLabelColors, 
  WhiteLabelTypography, 
  WhiteLabelLayout, 
  WhiteLabelBranding, 
  WhiteLabelKanban {}

// Dados do banco de dados (snake_case)
export interface WhiteLabelDBRecord {
  id: string;
  empresa_id: string;
  created_at?: string;
  updated_at?: string;
  // Branding
  title?: string;
  subtitle?: string;
  subject?: string;
  domain?: string;
  logo_url?: string;
  favicon_url?: string;
  login_image_url?: string;
  about_text?: string;
  email_footer?: string;
  empty_tone?: string;
  // Cores
  bg_color?: string;
  surface_color?: string;
  border_color?: string;
  text_color?: string;
  muted_color?: string;
  primary_color?: string;
  secondary_color?: string;
  link_color?: string;
  icon_color?: string;
  badge_bg?: string;
  success_color?: string;
  warning_color?: string;
  error_color?: string;
  info_color?: string;
  button_bg?: string;
  button_text?: string;
  button_hover?: string;
  button_disabled?: string;
  login_bg?: string;
  // Tipografia
  font_body?: string;
  font_heading?: string;
  base_font_size?: number;
  font_weight?: number;
  line_height?: number;
  // Layout
  density?: number;
  radius?: number;
  card_shadow?: number;
  // Kanban
  col_header_bg?: string;
  col_header_text?: string;
  col_border?: string;
  col_shadow?: string;
  col_width?: number;
  col_auto_width?: boolean;
  card_bg?: string;
  card_border?: string;
  card_stripe?: string;
  stripe_mode?: number;
  card_compact?: boolean;
  blocked_color?: string;
  f_title?: boolean;
  f_subtitle?: boolean;
  f_id?: boolean;
  f_tags?: boolean;
  f_assignee?: boolean;
  f_date?: boolean;
  f_sla?: boolean;
  f_priority?: boolean;
  f_points?: boolean;
  f_labels?: boolean;
  label_required?: boolean;
  label_limit?: number;
  label_palette?: string;
  avatar_shape?: string;
  avatar_size?: number;
  avatar_photo?: boolean;
  a_move?: boolean;
  a_done?: boolean;
  a_comment?: boolean;
  a_assign?: boolean;
}

// Constantes de CSS Variables
export const CSS_VARIABLES = {
  // Background
  BACKGROUND: '--background',
  CARD: '--card',
  POPOVER: '--popover',
  
  // Borders
  BORDER: '--border',
  INPUT: '--input',
  
  // Text
  FOREGROUND: '--foreground',
  CARD_FOREGROUND: '--card-foreground',
  POPOVER_FOREGROUND: '--popover-foreground',
  MUTED: '--muted',
  MUTED_FOREGROUND: '--muted-foreground',
  
  // Primary/Secondary
  PRIMARY: '--primary',
  SECONDARY: '--secondary',
  ACCENT: '--accent',
  RING: '--ring',
  
  // Sidebar
  SIDEBAR_BACKGROUND: '--sidebar-background',
  SIDEBAR_FOREGROUND: '--sidebar-foreground',
  SIDEBAR_PRIMARY: '--sidebar-primary',
  SIDEBAR_RING: '--sidebar-ring',
  SIDEBAR_ACCENT: '--sidebar-accent',
  SIDEBAR_ACCENT_FOREGROUND: '--sidebar-accent-foreground',
  SIDEBAR_BORDER: '--sidebar-border',
  
  // States
  SUCCESS: '--success',
  WARNING: '--warning',
  DESTRUCTIVE: '--destructive',
  INFO: '--info',
  
  // Layout
  RADIUS: '--radius',
  
  // Typography
  FONT_BODY: '--font-body',
  FONT_HEADING: '--font-heading',
  FONT_SIZE_BASE: '--font-size-base',
  FONT_WEIGHT_BASE: '--font-weight-base',
  LINE_HEIGHT_BASE: '--line-height-base',
} as const;

// Chaves do localStorage
export const STORAGE_KEYS = {
  CONFIG: 'wl_config',
  EMPRESA_SST_ID: 'wl_empresa_sst_id',
} as const;

File: apps/api/app/schemas/white_label.py
"""Schemas Pydantic para o módulo White Label / Configuração de empresa.

Tabelas cobertas:
- white_label_config  (1-per-empresa, tenant=empresa_id, UniqueConstraint)
- empresas_modulos    (many-per-empresa, tenant=empresa_id)
- empresas_modulos_telas (many-per-empresa, tenant=empresa_id)
- modulos             (global/read-only — sem empresa_id)
"""
import datetime
import uuid
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel


# ── Modulos (catálogo global — leitura para todos; escrita só admin_vertical) ──

class ModuloIn(BaseModel):
    nome: str
    rota: str
    descricao: Optional[str] = None
    icone: Optional[str] = None


class ModuloUpdate(BaseModel):
    nome: Optional[str] = None
    rota: Optional[str] = None
    descricao: Optional[str] = None
    icone: Optional[str] = None


class ModuloOut(BaseModel):
    id: uuid.UUID
    nome: str
    rota: str
    descricao: Optional[str] = None
    icone: Optional[str] = None
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


# ── EmpresasModulos ───────────────────────────────────────────────────────────

class EmpresaModuloIn(BaseModel):
    modulo_id: uuid.UUID
    ativo: bool = True


class EmpresaModuloUpdate(BaseModel):
    ativo: Optional[bool] = None


# ── Administração cross-tenant (super admin) — escopo pelo empresa_id do path ──
# Usados por /empresas/{empresa_id}/modulos*. Diferente do auto-serviço acima, o
# modulo_id vem do path; o corpo carrega só o estado.

class EmpresaModuloAtivoIn(BaseModel):
    ativo: bool = True


class TelasSetIn(BaseModel):
    """Conjunto EXATO de telas ativas de um módulo para a empresa.

    O servidor reconcilia: insere as ausentes, remove as que sumiram, reativa
    as presentes. Substitui o loop client-side de N requests por 1 chamada.
    """
    tela_ids: list[str]


class EmpresaModuloOut(BaseModel):
    id: uuid.UUID
    empresa_id: uuid.UUID
    modulo_id: uuid.UUID
    ativo: bool
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


# ── EmpresasModulosTelas ──────────────────────────────────────────────────────

class EmpresaModuloTelaIn(BaseModel):
    modulo_id: uuid.UUID
    tela_id: str
    ativo: Optional[bool] = True


class EmpresaModuloTelaUpdate(BaseModel):
    ativo: Optional[bool] = None


class EmpresaModuloTelaOut(BaseModel):
    id: uuid.UUID
    empresa_id: uuid.UUID
    modulo_id: uuid.UUID
    tela_id: str
    ativo: Optional[bool] = None
    created_at: Optional[datetime.datetime] = None
    updated_at: Optional[datetime.datetime] = None

    model_config = {"from_attributes": True}


# ── WhiteLabelConfig ──────────────────────────────────────────────────────────
# 1-per-empresa  →  GET (own config) + PUT (upsert own config)
# UPDATE schema excludes empresa_id (never allow tenant FK override).

class WhiteLabelConfigUpdate(BaseModel):
    title: Optional[str] = None
    subtitle: Optional[str] = None
    subject: Optional[str] = None
    domain: Optional[str] = None
    font_body: Optional[str] = None
    font_heading: Optional[str] = None
    base_font_size: Optional[int] = None
    font_weight: Optional[int] = None
    line_height: Optional[Decimal] = None
    density: Optional[Decimal] = None
    radius: Optional[int] = None
    card_shadow: Optional[Decimal] = None
    bg_color: Optional[str] = None
    surface_color: Optional[str] = None
    border_color: Optional[str] = None
    text_color: Optional[str] = None
    muted_color: Optional[str] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    link_color: Optional[str] = None
    icon_color: Optional[str] = None
    badge_bg: Optional[str] = None
    success_color: Optional[str] = None
    warning_color: Optional[str] = None
    error_color: Optional[str] = None
    info_color: Optional[str] = None
    button_bg: Optional[str] = None
    button_text: Optional[str] = None
    button_hover: Optional[str] = None
    button_disabled: Optional[str] = None
    empty_tone: Optional[str] = None
    login_bg: Optional[str] = None
    about_text: Optional[str] = None
    email_footer: Optional[str] = None
    logo_url: Optional[str] = None
    favicon_url: Optional[str] = None
    login_image_url: Optional[str] = None
    col_header_bg: Optional[str] = None
    col_header_text: Optional[str] = None
    col_border: Optional[str] = None
    col_shadow: Optional[str] = None
    col_width: Optional[int] = None
    col_auto_width: Optional[bool] = None
    card_bg: Optional[str] = None
    card_border: Optional[str] = None
    card_stripe: Optional[str] = None
    stripe_mode: Optional[int] = None
    card_compact: Optional[bool] = None
    blocked_color: Optional[str] = None
    f_title: Optional[bool] = None
    f_subtitle: Optional[bool] = None
    f_id: Optional[bool] = None
    f_tags: Optional[bool] = None
    f_assignee: Optional[bool] = None
    f_date: Optional[bool] = None
    f_sla: Optional[bool] = None
    f_priority: Optional[bool] = None
    f_points: Optional[bool] = None
    f_labels: Optional[bool] = None
    label_required: Optional[bool] = None
    label_limit: Optional[int] = None
    label_palette: Optional[str] = None
    avatar_shape: Optional[str] = None
    avatar_size: Optional[int] = None
    avatar_photo: Optional[bool] = None
    a_move: Optional[bool] = None
    a_done: Optional[bool] = None
    a_comment: Optional[bool] = None
    a_assign: Optional[bool] = None


class WhiteLabelConfigOut(BaseModel):
    id: uuid.UUID
    empresa_id: uuid.UUID
    title: Optional[str] = None
    subtitle: Optional[str] = None
    subject: Optional[str] = None
    domain: Optional[str] = None
    font_body: Optional[str] = None
    font_heading: Optional[str] = None
    base_font_size: Optional[int] = None
    font_weight: Optional[int] = None
    line_height: Optional[Decimal] = None
    density: Optional[Decimal] = None
    radius: Optional[int] = None
    card_shadow: Optional[Decimal] = None
    bg_color: Optional[str] = None
    surface_color: Optional[str] = None
    border_color: Optional[str] = None
    text_color: Optional[str] = None
    muted_color: Optional[str] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    link_color: Optional[str] = None
    icon_color: Optional[str] = None
    badge_bg: Optional[str] = None
    success_color: Optional[str] = None
    warning_color: Optional[str] = None
    error_color: Optional[str] = None
    info_color: Optional[str] = None
    button_bg: Optional[str] = None
    button_text: Optional[str] = None
    button_hover: Optional[str] = None
    button_disabled: Optional[str] = None
    empty_tone: Optional[str] = None
    login_bg: Optional[str] = None
    about_text: Optional[str] = None
    email_footer: Optional[str] = None
    logo_url: Optional[str] = None
    favicon_url: Optional[str] = None
    login_image_url: Optional[str] = None
    col_header_bg: Optional[str] = None
    col_header_text: Optional[str] = None
    col_border: Optional[str] = None
    col_shadow: Optional[str] = None
    col_width: Optional[int] = None
    col_auto_width: Optional[bool] = None
    card_bg: Optional[str] = None
    card_border: Optional[str] = None
    card_stripe: Optional[str] = None
    stripe_mode: Optional[int] = None
    card_compact: Optional[bool] = None
    blocked_color: Optional[str] = None
    f_title: Optional[bool] = None
    f_subtitle: Optional[bool] = None
    f_id: Optional[bool] = None
    f_tags: Optional[bool] = None
    f_assignee: Optional[bool] = None
    f_date: Optional[bool] = None
    f_sla: Optional[bool] = None
    f_priority: Optional[bool] = None
    f_points: Optional[bool] = None
    f_labels: Optional[bool] = None
    label_required: Optional[bool] = None
    label_limit: Optional[int] = None
    label_palette: Optional[str] = None
    avatar_shape: Optional[str] = None
    avatar_size: Optional[int] = None
    avatar_photo: Optional[bool] = None
    a_move: Optional[bool] = None
    a_done: Optional[bool] = None
    a_comment: Optional[bool] = None
    a_assign: Optional[bool] = None
    created_at: Optional[datetime.datetime] = None
    updated_at: Optional[datetime.datetime] = None

    model_config = {"from_attributes": True}
